plutus-core/testlib/PlutusCore/Generators/PIR/ShrinkTypes.hs

```diff
@@ -0,0 +1,252 @@
+--   Complies with the implicit shrinking order.
+--   TODO (later): also allow changing which context it's valid in
+fixKind :: HasCallStack
+        => TypeCtx
+        -> Type TyName DefaultUni ()
+        -> Kind ()
+        -> Type TyName DefaultUni ()
+fixKind ctx ty k
+  -- Nothing to do if we already have the right kind
+  | unsafeInferKind ctx ty == k = ty
+  | not $ k `leKind` unsafeInferKind ctx ty = error $ concat
+      [ "Internal error. New kind: "
+      , display k
+      , "\nis not smaller than the old one: "
+      , display $ unsafeInferKind ctx ty
+      ]
+  | otherwise = case ty of
+    -- Switch a variable out for a different variable of the right kind
+    TyVar _ _ -> case [y | (y, k') <- Map.toList ctx, k == k'] of
+        y : _ -> TyVar () y
+        _     -> minimalType k
+    -- Try to fix application by fixing the function
+    TyApp _ a b -> TyApp () (fixKind ctx a $ KindArrow () (unsafeInferKind ctx b) k) b
+    TyLam _ x kx b ->
+      case k of
+        -- Fix lambdas to * by substituting a minimal type for the argument
+        -- and fixing the body.
+        Type{} -> fixKind ctx (typeSubstClosedType x (minimalType kx) b) k
+        -- Fix functions by either keeping the argument around (if we can) or getting
+        -- rid of the argument (by turning its use-sites into minimal types) and introducing
+        -- a new argument.
+        KindArrow _ ka kb
+          | ka == kx              -> TyLam () x kx $ fixKind (Map.insert x kx ctx) b kb
+          | not $ kb `leKind` kb' -> error "fixKind"
+          | otherwise             -> TyLam () x ka $ fixKind ctx' b' kb
+            where
+              ctx' = Map.insert x ka ctx
+              b'   = typeSubstClosedType x (minimalType kx) b
+              kb'  = unsafeInferKind ctx' b'
+    -- Ill-kinded builtins just go to minimal types
+    TyBuiltin{} -> minimalType k
+    -- Unreachable, because there's no kind that is smaller than @Type ()@.
+    TyFun{} -> error "Internal error: unreachable clause."
+    TyIFix{} -> error "Internal error: unreachable clause."
+    TyForall{} -> error "Internal error: unreachable clause."
+
+-- | Shrink a well-kinded type in a context to new types, possibly with new kinds.
+-- The new kinds are guaranteed to be smaller than or equal to the old kind.
+-- TODO: also shrink to new context
+--       need old context and new context
+shrinkKindAndType :: HasCallStack
+                  => TypeCtx
+                  -> (Kind (), Type TyName DefaultUni ())
+                  -> [(Kind (), Type TyName DefaultUni ())]
+shrinkKindAndType ctx (k0, ty) =
+  -- If we are not already minimal, add the minial type as a possible shrink.
+  [(k, m) | k <- k0 : shrink k0, let m = minimalType k, m /= ty] ++
+  -- TODO: it might be worth-while to refactor this to the structural + nonstructural
+  -- style we use below. Unsure if that's more readable. CODE REVIEW: what do you think?
```

Comment:
I missed this comment, don't know what it means either. @MaximilianAlgehed do you remember?

Reply:
This refers to the style of shrinking function we wrote for terms where we split the "structural" cases (e.g. shrinking under application) from the "nonstructural" cases (i.e. removing whole subterms all together). The code for that shrinker is more readable than this code so it's probably a decent idea to rewrite this code in that style - but the comment obviously doesn't make much sense any more because it's pointing to code that's not in this PR.

Reply:
OK, I've looked into the term shrinker and I actually find the type shrinker more readable, because you can look at an AST node and observe all the different ways of shrinking it in one place instead of jumping between two, so I'm going to keep the current alignment.

I don't think I'll change the terms alignment either, 'cause it's also perfectly fine. I don't personally care about the slight inconsistency, both the shrinkers are fine.